monty-31/src/aarch64_neon/packing.rs

```diff
@@ -586,38 +591,261 @@ impl_packed_field_pow_2!(
+                    acc + general_dot_product::<_, 3>(&lhs_rem, &rhs_rem)
+                }
+                _ => unreachable!(),
+            }
+        }
+    }
+}
+
+/// Dot product of four packed columns with a single delayed reduction.
+#[inline]
+unsafe fn dot_product_4<P>(
+    lhs: &[MontyField31<P>; 4],
+    rhs: &[PackedMontyField31Neon<P>; 4],
+) -> PackedMontyField31Neon<P>
+where
+    P: MontyParametersNeon + FieldParameters,
+{
+    unsafe {
+        // Accumulate the full 64-bit sum C = Σ lhs_i ⋅ rhs_i.
+        //
+        // Start with lhs_0 ⋅ rhs_0 (split into low lanes 0,1 and high lanes 2,3).
+        let mut sum_l =
+            aarch64::vmull_n_u32(aarch64::vget_low_u32(rhs[0].to_vector()), lhs[0].value);
+        let mut sum_h = aarch64::vmull_high_n_u32(rhs[0].to_vector(), lhs[0].value);
+
+        // Add lhs_1 ⋅ rhs_1.
+        sum_l = aarch64::vmlal_n_u32(
+            sum_l,
+            aarch64::vget_low_u32(rhs[1].to_vector()),
+            lhs[1].value,
+        );
+        sum_h = aarch64::vmlal_high_n_u32(sum_h, rhs[1].to_vector(), lhs[1].value);
+
+        // Add lhs_2 ⋅ rhs_2.
+        sum_l = aarch64::vmlal_n_u32(
+            sum_l,
+            aarch64::vget_low_u32(rhs[2].to_vector()),
+            lhs[2].value,
+        );
+        sum_h = aarch64::vmlal_high_n_u32(sum_h, rhs[2].to_vector(), lhs[2].value);
+
+        // Add lhs_3 ⋅ rhs_3.
+        sum_l = aarch64::vmlal_n_u32(
+            sum_l,
+            aarch64::vget_low_u32(rhs[3].to_vector()),
+            lhs[3].value,
+        );
+        sum_h = aarch64::vmlal_high_n_u32(sum_h, rhs[3].to_vector(), lhs[3].value);
+
+        // Split C into 32-bit halves per lane:
+        // - c_lo = C mod 2^{32},
+        // - c_hi = C >> 32.
+        let c_lo = aarch64::vuzp1q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+        let c_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(sum_l),
+            aarch64::vreinterpretq_u32_u64(sum_h),
+        );
+
+        // Partial high-word reduction: c_hi' ∈ [0,P).
+        //
+        // Since C < 4P^2 and P < 2^{31}, we have c_hi < 2P.
+        //
+        // Map [P,2P) → [0,P) by subtracting P.
+        let c_hi_sub = aarch64::vsubq_u32(c_hi, P::PACKED_P);
+        let c_hi_prime = aarch64::vminq_u32(c_hi, c_hi_sub);
+
+        // q ≡ c_lo ⋅ μ (mod 2^{32}), with μ = −P^{-1} (mod 2^{32}).
+        let q = aarch64::vmulq_s32(aarch64::vreinterpretq_s32_u32(c_lo), P::PACKED_MU);
+        let q_u32 = aarch64::vreinterpretq_u32_s32(q);
+
+        // Compute (q⋅P)_hi = high 32 bits of q⋅P per lane (exact unsigned widening multiply).
+        let (qp_l, qp_h) = vmull_wide_u32(q_u32, P::PACKED_P);
+        let qp_hi = aarch64::vuzp2q_u32(
+            aarch64::vreinterpretq_u32_u64(qp_l),
+            aarch64::vreinterpretq_u32_u64(qp_h),
+        );
```

Comment:
Use `vqdmulhq_s32` to compute what we want (multiplied by 2).
```suggestion
        // Gets bits 31, ..., 62 of qp. Saturation is not an issue because `P` is not -2**31.
        let qp_2_hi = aarch64::vqdmulhq_s32(q, aarch64::vreinterpretq_s32_u32(MPNeon::PACKED_P));
        let qp_hi = aarch64::vshrq_n_u32::<1>(qp_2_hi);
```

Reply:
For some reason this trick seems not to work. I think I get the idea but with the current function seems not to work properly